gcp_variant_transforms/transforms/extract_variant_header.py

```diff
@@ -0,0 +1,137 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A PTransform to output a PCollection of single ``HeaderFields`` extracted
+from ``Variant`` recoreds. This ``HeaderFields`` contains only fields (FORMAT,
+INFO) not already defined in the header section of the VCF files."""
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+
+from vcf.parser import _Format as Format
+from vcf.parser import _Info as Info
+from vcf.parser import field_counts
+
+from gcp_variant_transforms.beam_io import vcf_header_io
+from gcp_variant_transforms.transforms import merge_headers
+
+
+__all__ = ['ExtractHeaderFieldsFromVariant']
+
+
+_FLAG = 'Flag'
+_STRING = 'String'
+_NO_DESCRIPTION = 'no_description'
+_UNKNOWN_SOURCE = 'unknown_source'
+_UNKNOWN_VERSION = 'unknown_version'
+
+
+class _ExtractHeaderFieldsFromVariant(beam.DoFn):
+  """Extracts header fields from ``Variant`` record."""
+
+  def _is_defined_info_field(self, field_key, headers):
+    return headers and field_key in headers.infos
+
+
+  def _is_defined_format_field(self, field_key, headers):
+    return headers and field_key in headers.formats
+
+
+  def _get_field_counts(self, field_value):
+    # TODO(nmousavi): return filed count by conditioning on field_value type
+    # (list, bool, o/w) once variant_to_bigquery code can deduct type for
+    # missing header fields from extracted final headers. Currently it inserts
+    # fields with missing header to bigquery as a repeated field.
+    return field_counts['.']
+
+
+  def _get_field_type(self, field_value):
+    if isinstance(field_value, bool):
+      return _FLAG
+    else:
+      return _STRING
+
+
+  def _extract_info_fields(self, variant, headers):
+    infos = {}
+    for info_field_key, variant_info in variant.info.iteritems():
+      info_field_value = variant_info.data
+      if not self._is_defined_info_field(info_field_key, headers):
+        assert info_field_key not in infos, (
+            "Bad VCF file. Duplicate INFO field. %s" %info_field_key)
+        infos[info_field_key] = Info(info_field_key,
+                                     self._get_field_counts(info_field_value),
+                                     self._get_field_type(info_field_value),
+                                     _NO_DESCRIPTION,
+                                     _UNKNOWN_SOURCE,
+                                     _UNKNOWN_VERSION)
+    return infos
+
+
+  def _extract_format_fields(self, variant, headers):
+    formats = {}
+    for call in variant.calls:
+      for format_key, format_value in call.info.iteritems():
+        if not self._is_defined_format_field(format_key, headers):
+          assert format_key not in formats, (
```

Comment:
Consider raising a ValueError instead of asserting (e.g. we may want to catch the error and do something with it later instead of crashing).

Reply:
Done